### ISSUE TITLE
JVM is crashed on strange scala code

### ISSUE THREAD

**Issue by catap** [AdoptOpenJDK/openjdk-support#207]
I do have some project and accidentally I discovered that it crashed JVM.

Code that works:
```
  private def doTransfer(): Future[Unit] = {
    Future {
      val fulfilledObligations = new mutable.Queue[(Array[Byte], Promise[Array[Byte]])]()

      synchronized {
        while (available.nonEmpty && promises.nonEmpty) {
          // some logic
        }
      }

      while (fulfilledObligations.nonEmpty) {
        val (bytes, promise) = fulfilledObligations.dequeue()
        promise.trySuccess(bytes)
      }
    }
  }
```
if I remove `{` and `}` from beginning and at the end of this function, like this:
```
  private def doTransfer(): Future[Unit] =
    Future {
      val fulfilledObligations = new mutable.Queue[(Array[Byte], Promise[Array[Byte]])]()

      synchronized {
        while (available.nonEmpty && promises.nonEmpty) {
          // some logic
        }
      }

      while (fulfilledObligations.nonEmpty) {
        val (bytes, promise) = fulfilledObligations.dequeue()
        promise.trySuccess(bytes)
      }
    }
```

... it's crashed with dump:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000110caf908, pid=64459, tid=27907
#
# JRE version: OpenJDK Runtime Environment AdoptOpenJDK (11.0.8+10) (build 11.0.8+10)
# Java VM: OpenJDK 64-Bit Server VM AdoptOpenJDK (11.0.8+10, mixed mode, tiered, compressed oops, g1 gc, bsd-amd64)
# Problematic frame:
# V  [libjvm.dylib+0x6af908]  SubINode::Ideal(PhaseGVN*, bool)+0x22c
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# If you would like to submit a bug report, please visit:
#   https://github.com/AdoptOpenJDK/openjdk-support/issues
#

---------------  S U M M A R Y ------------

Command Line: -javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=54164:/Applications/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8 org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner @/var/folders/rq/wqy42lq543z9rwrlj7vf1h4r0000gn/T/idea_scala_test_runner14727095344342102070.tmp -showProgressMessages true

Host: MacBookPro15,1 x86_64 2900 MHz, 12 cores, 32G, Darwin 19.6.0
Time: Tue Nov 17 18:21:35 2020 CET elapsed time: 3 seconds (0d 0h 0m 3s)

---------------  T H R E A D  ---------------

Current thread (0x00007fd91e17b000):  JavaThread "C2 CompilerThread2" daemon [_thread_in_native, id=27907, stack(0x00007000104b1000,0x00007000105b1000)]


Current CompileTask:
C2:   3541 4782   !   4       xxx.BufferedAsyncPipe::$anonfun$doTransfer$1 (429 bytes)

Stack: [0x00007000104b1000,0x00007000105b1000],  sp=0x00007000105adc00,  free space=1011k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.dylib+0x6af908]  SubINode::Ideal(PhaseGVN*, bool)+0x22c
V  [libjvm.dylib+0x627d5d]  PhaseIterGVN::transform_old(Node*)+0x5d
V  [libjvm.dylib+0x37ba46]  IfNode::fold_compares_helper(ProjNode*, ProjNode*, ProjNode*, PhaseIterGVN*)+0x860
V  [libjvm.dylib+0x37c66b]  IfNode::fold_compares(PhaseIterGVN*)+0xd9
V  [libjvm.dylib+0x37daa8]  IfNode::Ideal(PhaseGVN*, bool)+0xcc
V  [libjvm.dylib+0x627d5d]  PhaseIterGVN::transform_old(Node*)+0x5d
V  [libjvm.dylib+0x627b9e]  PhaseIterGVN::optimize()+0x86
V  [libjvm.dylib+0x21e3fa]  Compile::Optimize()+0xbca
V  [libjvm.dylib+0x21cb8f]  Compile::Compile(ciEnv*, C2Compiler*, ciMethod*, int, bool, bool, bool, DirectiveSet*)+0x853
V  [libjvm.dylib+0x19e270]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, DirectiveSet*)+0xb6
V  [libjvm.dylib+0x229567]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0x577
V  [libjvm.dylib+0x228e7b]  CompileBroker::compiler_thread_loop()+0x1ab
V  [libjvm.dylib+0x6ec6e0]  JavaThread::thread_main_inner()+0x82
V  [libjvm.dylib+0x6ec52a]  JavaThread::run()+0x174
V  [libjvm.dylib+0x6ea406]  Thread::call_run()+0x68
V  [libjvm.dylib+0x5f2233]  thread_native_entry(Thread*)+0x139
C  [libsystem_pthread.dylib+0x6109]  _pthread_start+0x94
C  [libsystem_pthread.dylib+0x1b8b]  thread_start+0xf


siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x0000000000000028
```

This code was compiled by stock scala-2.13.3

**Comment by aahlenst** [comment]
Can you please check with 11.0.9.1 from https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/tag/jdk-11.0.9.1%2B1?

**Comment by catap** [comment]
@aahlenst I can't reproduce it on 11.0.9.1

I'm curious: was it a blind shot or you remember some issue that might be related?

**Comment by karianna** [comment]
There was a critical C2 compiler bug fix in 11.0.9.1

**Comment by aahlenst** [comment]
@catap Martijn is probably not correct. The C2 bug he mentions was only in 11.0.9, but your crash dump is from 11.0.8. My hunch was it could be https://bugs.openjdk.java.net/browse/JDK-8250609 which was fixed with 11.0.9 although it's not an exact match.

**Comment by catap** [comment]
@aahlenst I found the exact issue that I see here: https://bugs.openjdk.java.net/browse/JDK-8250861
